### PR TITLE
Kernel+Userland: Introduce containers

### PR DESCRIPTION
**Pull request by supercomputer7**
This is the thing I worked on for the last 2 weeks - containers, which consist of combining 4 main features:
- Jail restrictions
- PID isolation in Process lists
- VFS root contexts
- Hostname contexts

The concept of VFS root context is very similar to the Linux mount namespace idea. I still have to write all the reasoning behind what I do, so this is a draft, for now.
In the next couple of days I will try to land the last commits of bringing it all up with a new shiny userspace utility called `runc` (for "run container").

Please note that although this is not finished yet, there are many good bits in the PR which will benefit the system in so many ways, and because I use them as foundations for the end goal, I rather keep them in this PR.

### CODE REVIEW

**Comment by supercomputer7** [comment]
Now this is mostly done in terms of finishing the design, I still need to fix CI and ensure proper commit messages are in place.

As for future development, we should aim for adding sysfs nodes for unshared resources. This will probably require us to also keep a list of attached process to a resource (like with the `ScopedProcessList` class) if we want to do something meaningful.
In addition to that, I searched on how we could access VFS root context filesystem view, and this serverfault [question](https://serverfault.com/questions/878783/access-another-processs-mount-namespace) has some interesting answers.

**Comment by supercomputer7** [comment]
More ideas that came to my mind:
- Adding separation to SlavePTY devices so we can mount `devpts` filesystems within a container, and based on the VFS root context (which could hold a list instead of the global list), the processes can share a new set of devices contained in their own filesystem instance in such VFS root context.
- Instantiate separation for other devices so maybe attach each device to certain VFS root context so it could be opened only within the same VFS root context. If you open a device node before moving to other VFS root context, then you keep that device node being opened in a file descriptor. This will allow creating device nodes without the fear of processes trying to open them to bypass the container isolation (so you can't open the main disk device node and read from it).

**Comment by supercomputer7** [comment]
Another clever idea - we could take leverage of the VFS root context idea to implement lazily-unmount feature - essentially we can allow the user to ask the kernel to move the filesystem out of its context, attaching it to the kernel VFS root context which is not visible for userspace, and then add some kernel process worker that will clean it up once the last references to the filesystem inodes are cleaned.

**Comment by supercomputer7** [comment]
This is blocked until #23138 is merged, since I plan on expanding on loop devices for containers as well.

**Comment by OfficialPixelBrush** [comment]
Any updates on this front?

**Comment by supercomputer7** [comment]
I noticed that now - I should probably add a manual page for the `runc` utility soon. Still, suggestions and reviews are welcomed! :)

**Comment by supercomputer7** [comment]
Since I added the SetOnce flag in #24071, we need to back-port that change to this PR when it get merged.

**Comment by supercomputer7** [comment]
I ported https://github.com/supercomputer7/serenity/commit/3eefaf323e02b7d3162cb8ad97db61152a0f7301 commit from #24071 (without converting other use-cases as they're not relevant to this PR), so we could merge this as-is, without relying on the other PR :)

**Comment by supercomputer7** [comment]
Just putting another idea here as a note - to implement a way to auto-jail processes when starting suitable programs for such policy. Either through a special ELF flag, or a separate dynamic loader which uses hardened policy.

**Comment by supercomputer7** [comment]
CI is flaky a bit, but the `fusermount` utility needs to be updated so it will use the slightly-changed `Core::System` mount API.

**Comment by supercomputer7** [comment]
Just putting another idea here as a note - to implement a way to manage containers, something like `containerd` on Linux.
Then `runc` can connect to a local `unix` socket of such service and ask the service to launch or stop containers, and probably manage graceful tear-down of a container as well.
This way we could use loop devices and manage pre-built images and mount them in new containers, like how it works in `docker` for example.

File: Kernel/Security/Jail.cpp
Comment (timschumi) [inline Kernel/Security/Jail.cpp:28]: I'm sure you had your reasons for this (and neither am I arguing in favor or against in particular), but doing a 180 on the whole feature like this is not exactly confidence-inspiring.

File: Kernel/API/Syscall.h
Comment (timschumi) [inline Kernel/API/Syscall.h:140]: We've tried this before I think, although I don't remember the exact reasoning why we ultimately abandoned the idea.

I myself am even questioning whether "I only need \<list\> going forward" or "I don't need \<list\> going forward" is the weirder pattern to have. It's probably more fool-proof to have the application developer keep in mind which capabilities their program is effectively keeping.

File: Userland/Utilities/copy_mount.cpp
Comment (timschumi) [inline Userland/Utilities/copy_mount.cpp:29]: Looks like those options are `ArgsParser::Required` after all?

File: Kernel/Tasks/HostnameContext.cpp
Comment (timschumi) [inline Kernel/Tasks/HostnameContext.cpp:68]: Should we move the condition outside to avoid acquiring the lock?

File: Userland/BuggieBox/main.cpp
Comment (timschumi) [inline Userland/BuggieBox/main.cpp:115]: I'd argue that this is not really BuggieBox' duty. If we can call BuggieBox to run `symlinks`, we can just use the commands directly.

File: Userland/BuggieBox/main.cpp
Comment (timschumi) [inline Userland/BuggieBox/main.cpp:108]: Can't we just substitute this with `strings[0]`?

File: Userland/Libraries/LibCore/System.h
Comment (timschumi) [inline Userland/Libraries/LibCore/System.h:214]: Probably nicer for searchability if these are `unshare_create` and `unshare_attach` respectively.

File: Base/usr/share/runc/full-buggiebox-container.json
Comment (timschumi) [inline Base/usr/share/runc/full-buggiebox-container.json:28]: Why don't we put poor BuggieBox in `/bin`?

File: Userland/Utilities/runc/main.cpp
Comment (timschumi) [inline Userland/Utilities/runc/main.cpp:39]: Also not really a NOTE, at least not with that formulation.

**Review by timschumi — CHANGES_REQUESTED** [review]
Not too bad structurally, just the usual amount of nitpicks, clarifications, etc. Building also currently fails.

File: Kernel/Security/Jail.cpp
Comment (supercomputer7) [reply Kernel/Security/Jail.cpp:28]: I just realized that making everything tied to one object in the kernel is not the right solution, as it's not flexible enough. By introducing what is essentially our own implementation of Linux `namespaces`, we can achieve full isolation in a reasonable fashion and configuration of such container can happen incrementally instead of one syscall (`create_jail`) that you can't actually reason about its return value.

File: Kernel/API/Syscall.h
Comment (supercomputer7) [reply Kernel/API/Syscall.h:140]: We had a whole discussion [here](https://github.com/SerenityOS/serenity/pull/23389). But sure, doing this in userspace is more smart. We could have LibCore API to actually subtract the pledges if we want to.

File: Userland/Libraries/LibC/unistd.h
Comment (supercomputer7) [reply Userland/Libraries/LibC/unistd.h:125]: So we expect libc API to be stable here? I don’t really mind it, if we can use LibCore and ensure libc just passes -1 as the vfs root context id.

Comment (timschumi) [reply Userland/Libraries/LibC/unistd.h:125]: > So we expect libc API to be stable here? 

At the very least `mount` and `umount` are standardized, so yes.

File: Kernel/Security/Jail.cpp
Comment (supercomputer7) [reply Kernel/Security/Jail.cpp:28]: Also, I'd like to put a note that jails are not entirely removed. It's just changing its purpose - it sets a new goal for it, so being jailed now is simply a kernel flag which restricts you from doing many things just like before (mainly dangerous syscalls but not just that). The other restrictions like PID isolation are moved elsewhere, because I'd like to build a flexible solution.

**Comment by supercomputer7** [comment]
This should be ready for review again :)

**Review by timschumi — APPROVED** [review]
Let's get this conflict magnet out of the way.

**Pull request merged**
